### PR TITLE
Request operations

### PR DESCRIPTION
**Pull request by bretambrose**
* Support for request-response operations
* Refactoring of rr-client <-> subscription manager contract/responsibilities to support requests without correlation tokens (fleet provisioning service)
* tests

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/request-response/request_response_client.c [hunk 1]
```diff
@@ -92,22 +105,17 @@ Client Tables/Lookups
     (Authoritative operation container)
```
Comment (sbSteveK) [inline source/request-response/request_response_client.c:105]: Trivial: This comment block is super helpful for quick understanding of the 4 current hash tables. They're easily identifiable but could we still label these with the names used for the members in the rr client below?

File: source/request-response/request_response_client.c [hunk 2]
```diff
@@ -913,6 +887,139 @@ static void s_apply_publish_to_streaming_operation_list(
...
+    if (operation->state == AWS_MRROS_PENDING_DESTROY) {
+        AWS_LOGF_DEBUG(
+            AWS_LS_MQTT_REQUEST_RESPONSE,
+            "id=%p: request-response operation %" PRIu64 " cannot be completed, already in pending destruction state",
+            (void *)operation->client_internal_ref,
+            operation->id);
+        return;
+    }
+
+    AWS_LOGF_DEBUG(
```
Comment (sbSteveK) [inline source/request-response/request_response_client.c:923]: Trivial: Should this log occur if the completion callback is invoked? Is this a failed operation if there's no completion callback and the payload of the correlated publish event isn't delivered anywhere?

**Review by sbSteveK — APPROVED** [review]

File: source/request-response/request_response_client.c (hunk 2 above)
Comment (bretambrose) [reply source/request-response/request_response_client.c:923]: If it's in pending destroy something else already "successfully" completed it (like a timeout).

File: source/request-response/request_response_client.c [hunk 3]
```diff
@@ -1187,14 +1323,68 @@ static int s_add_operation_to_subscription_topic_filter_table(
...
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_add_request_operation_to_correlation_token_table(
+    struct aws_mqtt_request_response_client *client,
+    struct aws_mqtt_rr_client_operation *operation) {
 
-    // TODO: NYI
+    return aws_hash_table_put(
```
[the remark this answers is not in the document]
Comment (bretambrose) [reply source/request-response/request_response_client.c:1361]: Duplicates are checked before dequeing.  If the head of the queue is in conflict with an in-progress operation, the queue is blocked until the operation completes.

File: source/request-response/request_response_client.c (hunk 1 above)
Comment (bretambrose) [reply source/request-response/request_response_client.c:105]: Updated in FinalPolish PR

**Pull request closed without merging**